src/main/java/com/based/basedsurvey/controller/FillSurveyController.java

```diff
@@ -54,29 +60,87 @@ public String getAnswer(@PathVariable Long surveyID, Model model) {
+                formInputs.append(getOpenAnswerInput((OpenAnswerQuestion) q));
 
                 // case 3, the question is a range question
             } else if (q instanceof RangeQuestion) {
-                formInputs += getRangeInput((RangeQuestion) q);
+                formInputs.append(getRangeInput((RangeQuestion) q));
             }
         }
 
-        model.addAttribute("questions", formInputs);
+        model.addAttribute("questions", formInputs.toString());
         return "Answer";
     }
 
+    @GetMapping("/survey/{surveyID}/answer/{page}")
+    @ResponseBody
+    public String getAnswer(@PathVariable Long surveyID, @PathVariable int page, Model model) {
+        // if survey does not exist
+        if (!surveyRepository.existsById(surveyID)) {
+            model.addAttribute("issue", "Survey does not exist");
+            return "SurveyFillIssue";
+        }
+
+        model.addAttribute("surveyID", surveyID);
+        Survey s = this.surveyRepository.findSurveyById(surveyID);
+        model.addAttribute("surveyName", s.getName());
+
+        // if survey is still being edited
+        if (s.getStatus() == Survey.SurveyStatuses.BEING_EDITED) {
+            model.addAttribute("issue", "Survey is being edited");
+            return "SurveyFillIssue";
+        }
+
+        // if survey is still finished
+        if (s.getStatus() == Survey.SurveyStatuses.FINISHED) {
+            model.addAttribute("issue", "Survey is finished");
+            return "SurveyFillIssue";
+        }
+
+        // if survey has no questions
+        if (s.getQuestions().size() == 0) {
+            model.addAttribute("issue", "Survey has no questions");
+            return "SurveyFillIssue";
+        }
+
+        // formInputs will contain a dynamically generated string with all the html form inputs
+        StringBuilder formInputs = new StringBuilder();
+        PageRequest pageRequest = PageRequest.of(page, pageSize, Sort.by("id").ascending());
+
+        var questions = questionRepository.findAllBySurveyId(surveyID, pageRequest).getContent();
+        if (questions.isEmpty()) return ""; // to prevent infinite requests from pages that have 0 elements
+
+        // for each question in survey
+        for (Question q : questions) {
+
+            // case 1, the question is a multiple choice question
+            if (q instanceof MultipleChoiceQuestion) {
+                formInputs.append(getMultipleChoiceInputs((MultipleChoiceQuestion) q));
+
+                // case 2, the question is an open answer question
+            } else if (q instanceof OpenAnswerQuestion) {
+                formInputs.append(getOpenAnswerInput((OpenAnswerQuestion) q));
+
+                // case 3, the question is a range question
+            } else if (q instanceof RangeQuestion) {
+                formInputs.append(getRangeInput((RangeQuestion) q));
+            }
+        }
+        formInputs.append(getPaginateHtml(surveyID, page + 1));
```

Comment:
does this work even if no pages are left?

Reply:
yes there is a check earlier if paginated query is empty